lib/Transforms/YosysOptimizer/YosysOptimizer.cpp

```diff
@@ -79,63 +82,234 @@ struct YosysOptimizer : public impl::YosysOptimizerBase<YosysOptimizer> {
+  typeConvertedArgs.reserve(op->getNumOperands());
+  if (failed(convertOpOperands(op, func, typeConvertedArgs))) {
+    return failure();
+  }
 
-    // Invoke Yosys to translate to a combinational circuit and optimize.
-    Yosys::yosys_setup();
-    Yosys::log_error_stderr = true;
-    LLVM_DEBUG(Yosys::log_streams.push_back(&std::cout));
-    Yosys::run_pass(llvm::formatv(kYosysTemplate.data(), filename,
-                                  op.getSymName(), yosysFilesPath, abcPath,
-                                  abcFast ? "-fast" : ""));
-
-    // Translate to MLIR and insert into the func
-    LLVM_DEBUG(Yosys::run_pass("dump;"));
-    std::stringstream cellOrder;
-    Yosys::log_streams.push_back(&cellOrder);
-    Yosys::run_pass("torder -stop * P*;");
-    Yosys::log_streams.clear();
-    auto topologicalOrder = getTopologicalOrder(cellOrder);
-
-    // Insert the optimized MLIR.
-    LUTImporter lutImporter = LUTImporter(&getContext());
-    Yosys::RTLIL::Design *design = Yosys::yosys_get_design();
-    func::FuncOp func =
-        lutImporter.importModule(design->top_module(), topologicalOrder);
-
-    LLVM_DEBUG(llvm::dbgs()
-               << "Converted & optimized func via yosys. Input func:\n"
-               << op << "\n\nOutput func:\n"
-               << func << "\n");
-    op.setFunctionType(func.getFunctionType());
-    op.getBody().takeBody(func.getBody());
+  int resultIndex = 0;
+  for (Type ty : func.getFunctionType().getResults())
+    op->getResult(resultIndex++).setType(secret::SecretType::get(ty));
 
+  // Replace the func.return with a secret.yield
+  op.getRegion().takeBody(func.getBody());
+  op.getOperation()->setOperands(typeConvertedArgs);
+
+  Block &block = op.getRegion().getBlocks().front();
+  func::ReturnOp returnOp = cast<func::ReturnOp>(block.getTerminator());
+  OpBuilder bodyBuilder(&block, block.end());
+  bodyBuilder.create<secret::YieldOp>(returnOp.getLoc(),
+                                      returnOp.getOperands());
+  returnOp.erase();
+  func.erase();
+
+  DenseSet<Operation *> castOps;
+  SmallVector<Value> typeConvertedResults;
+  castOps.reserve(op->getNumResults());
+  typeConvertedResults.reserve(op->getNumResults());
+  if (failed(convertOpResults(op, castOps, typeConvertedResults))) {
+    return failure();
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "Generic results: " << typeConvertedResults.size()
+                          << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "Original results: " << op.getResults().size()
+                          << "\n");
+
+  op.getResults().replaceUsesWithIf(
+      typeConvertedResults, [&](OpOperand &operand) {
+        return !castOps.contains(operand.getOwner());
+      });
+  return success();
+}
+
+// Optimize the body of a secret.generic op.
+// FIXME: consider utilizing
+// https://mlir.llvm.org/docs/PassManagement/#dynamic-pass-pipelines
+void YosysOptimizer::runOnOperation() {
+  auto result = getOperation()->walk([&](secret::GenericOp op) {
```

Comment:
Took me a while to realize, but right now we only end up caring about secret.generic bodies (since all out args are ciphertext args). later, lets say we have some plaintext value arg and there's some circuit logic on that arg outside of a secret.generic - in that case, we still operate on the plaintext value in normal plaintext arithmetic, but it may need to be converted to a booleanized input depending on the case - we tackle that later.

Reply:
So this pass still does have the `arith` ops required to booleanify plaintext inputs to the generic. That's in the branch on line 120 calling to `convertIntegerValue`. I opted for that instead of using something like `unrealized_conversion_cast`, but let me know if you think I should tweak it.

```c++
    if (!opOperand.get().getType().isa<secret::SecretType>()) {
      // The type is not secret, but still must be booleanized
      OpBuilder builder(op);
      auto fromElementsOp = convertIntegerValue(opOperand.get(), convertedType,
                                                builder, op.getLoc());
      typeConvertedArgs.push_back(fromElementsOp.getResult());
      continue;
    }
```

This implies that the circuit logic has to be inside the generic, and it seems that we would be better to fold it to occur _outside_ the generic (filed https://github.com/google/heir/issues/299 for a canonicalization pattern to hoist ops that can be hoisted)

Reply:
Actually the case I was thinking of was something like this:

```
func @main(%plaintext: i8, %secret_data: i8) {
   %0 = arith.addi 100, %plaintext
   secret.generic(%0, %secret_data) { ... }
} 
```

so that first instruction happens before the booleanization for secret.generic.

In this case, we skip booleanizing that first instruction at all, but I think that's fine for now (or maybe that's what you mean by " it seems that we would be better to fold it to occur outside the generic"?).